src/storage-client/src/controller.rs

```diff
@@ -3168,44 +3170,83 @@ where
+    ///       since is <= `f`.
+    ///
+    ///     - Its write frontier is neither `T::minimum` nor beyond the
+    ///       dependency since.
+    /// - If any identified collection's data source is not
+    ///   [`DataSource::Ingestion] (primary source) or [`DataSource::Other`]
+    ///   (subsources).
     fn install_dependency_read_holds<I: Iterator<Item = GlobalId>>(
         &mut self,
         collections: I,
         storage_dependencies: &[GlobalId],
     ) -> Result<(), StorageError> {
         let dependency_since = self.determine_collection_since_joins(storage_dependencies)?;
+
         for id in collections {
             let collection = self.collection(id).expect("known to exist");
+            assert!(
+                matches!(collection.description.data_source, DataSource::Other | DataSource::Ingestion(_)),
+                "only primary sources w/ subsources and subsources can have dependency read holds installed"
+            );
 
-            // At the time of collection creation, we did not yet
-            // have firm guarantees that the since of our
-            // dependencies was not advanced beyond those of its
-            // dependents, so we need to patch up the
-            // implied_capability/since of the collction.
+            // Because of the "backward" dependency structure (primary sources
+            // depend on subsources, rather than the other way around, which one
+            // might expect), we do not know what the initial since of the
+            // collection should be. We only find out that information once its
+            // primary sources comes along and correlates the subsource to its
+            // dependency sinces (e.g. remap shards).
+            //
+            // Once we find that out, we need ensure that the controller's
+            // version of the since is sufficiently advanced so that we may
+            // install the read hold.
             //
-            // TODO(aljoscha): This comes largely from the fact that
-            // subsources are created with a `DataSource::Other`, so
-            // we have no idea (at their creation time) that they
-            // are a subsource, or that they are a subsource of a
-            // source where they need a read hold on that
-            // ingestion's remap collection.
+            // TODO: remove this if statement once we fix the inverse dependency
+            // of subsources
             if PartialOrder::less_than(&collection.implied_capability, &dependency_since) {
                 assert!(
-                    PartialOrder::less_than(&dependency_since, &collection.write_frontier),
-                    "write frontier ({:?}) must be in advance dependency collection's since ({:?})",
+                    // This assertion is meant to mirror
+                    // https://github.com/MaterializeInc/materialize/blob/b997fa24db04bfaf4465421e814ceff167543fa0/src/storage/src/source/reclock/compat.rs#L117,
```

Comment:
Removed the comments but left the assert because if it ever occurred having the GlobalId and the fact that it's the remap shard would help debug it; removing it also makes testing the failpoint above it less obvious and in need of documentation.